### PR TITLE
[Core] Document TokenCredential implementation requirements

### PR DESCRIPTION
**Pull request by mccoyp**
# Description

Resolves https://github.com/Azure/azure-sdk-for-python/issues/22853 and resolves https://github.com/Azure/azure-sdk-for-python/issues/22824. This adds a section to the client library developer doc, describing the TokenCredential protocol and the requirements for implementing such a credential. The current plan is to link to the "TokenCredential protocol" section anchor with an `aka.ms` link: https://aka.ms/azsdk/python/identity/tokencredential. I'm open to suggestions for a different URL!

This also updates the TokenCredential protocol in Core to reflect changes in the past year to accept `claims` and `tenant_id`. A section of the TokenCredential documentation covers the current known uses of these parameters for reference and record keeping.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- _N/A_ Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/core/azure-core/azure/core/credentials.py [hunk 1]
```diff
@@ -8,21 +8,25 @@
...
     AccessToken = NamedTuple("AccessToken", [("token", str), ("expires_on", int)])
 
     class TokenCredential(Protocol):
-        """Protocol for classes able to provide OAuth tokens.
+        """Protocol for classes able to provide OAuth tokens."""
 
-        :param str scopes: Lets you specify the type of access needed.
-        """
+        def get_token(
+            self, *scopes: str, claims: Optional[str] = None, tenant_id: Optional[str] = None, **kwargs: Any
```
Comment (xiangyan99) [inline sdk/core/azure-core/azure/core/credentials.py:20]: not 
```py
self, *scopes: str, *, claims: Optional[str] = None, tenant_id: Optional[str] = None, **kwargs: Any
```
?

Comment (mccoyp) [reply sdk/core/azure-core/azure/core/credentials.py:20]: Laurent and I also just learned about this! The `*` argument isn't allowed after `*args` (or `*scopes`, in this case) because `*args` will capture any non-keyword arguments, making `*` unnecessary.

**Review by xiangyan99 — APPROVED** [review]

File: sdk/core/azure-core/CLIENT_LIBRARY_DEVELOPER.md [hunk 2]
```diff
@@ -522,3 +522,58 @@ class Pipeline:
         return first_node.send(pipeline_request)  # type: ignore
 
 ```
+
+## Credentials
+
+### TokenCredential protocol
+
+Clients from the Azure SDK often require a `TokenCredential` instance in their constructors. A `TokenCredential` is
```
Comment (lmazuel) [inline sdk/core/azure-core/CLIENT_LIBRARY_DEVELOPER.md:530]: I could be picky and say that clients that needs a AAD token require TokenCredential. We have SDKs that do not require it.

File: sdk/core/azure-core/CLIENT_LIBRARY_DEVELOPER.md [hunk 3]
```diff
@@ -522,3 +522,58 @@ class Pipeline:
...
+meant to provide OAuth tokens to authenticate service requests and can be implemented in a number of ways.
+
+The `TokenCredential` protocol specifies a class that has a single method -- `get_token` -- which returns an
+`AccessToken`: a `NamedTuple` containing a `token` string and an `expires_on` integer.
```
Comment (lmazuel) [inline sdk/core/azure-core/CLIENT_LIBRARY_DEVELOPER.md:534]: Can I get a unit for expires_on? Seconds? ms?

File: sdk/core/azure-core/CLIENT_LIBRARY_DEVELOPER.md (hunk 2 above)
Comment (mccoyp) [reply sdk/core/azure-core/CLIENT_LIBRARY_DEVELOPER.md:530]: Would it be more clear to rephrase this to "Most clients from the Azure SDK require a `TokenCredential` instance in their constructors"?

Comment (lmazuel) [reply sdk/core/azure-core/CLIENT_LIBRARY_DEVELOPER.md:530]: Well, re-reading I'm fine, let's solve the discussion

File: sdk/core/azure-core/azure/core/credentials_async.py [hunk 4]
```diff
@@ -5,13 +5,24 @@
...
+            self, *scopes: str, claims: Optional[str] = None, tenant_id: Optional[str] = None, **kwargs: Any
+        ) -> AccessToken:
+            """Request an access token for `scopes`.
+
+            :param str scopes: The type of access needed.
+
+            :keyword str claims: Additional claims required in the token, such as those returned in a resource
+                provider's claims challenge following an authorization failure.
+            :keyword str tenant_id: Optional tenant to include in the token request.
+            """
```
Comment (lmazuel) [inline sdk/core/azure-core/azure/core/credentials_async.py:25]:
```python
:rtype: AccessToken
:return: An instance of AccessToken containing the token itself and the expiration time
```

**Review by lmazuel — APPROVED** [review]
